### ISSUE TITLE
Aggregation Service: GROUP BY support

### ISSUE THREAD

**Issue by JonathanAquino-NextRoll** [WICG/attribution-reporting-api#256]
We often need to report on a high cardinality attribute like campaign id. For ease of reporting and smaller query strings, it would be great if we could use something like GROUP BY to retrieve all of the different values of the attribute, instead of having to specify each value explicitly.

Something like:

    select campaign_id, sum(impressions), sum(spend)
    where date >= 2021-09-01 and date < 2021-10-01
    group by campaign_id

Grouping by multiple attributes:

    select campaign_id, foo, sum(impressions), sum(spend)
    where date >= 2021-09-01 and date < 2021-10-01
    group by campaign_id, foo

**Comment by csharrison** [comment]
Thanks this is something we are considering, but we need to weigh it with the increase in complexity for the servers. The current design of a fixed histogram allows a lot of flexibility with how the computation can be performed.

**Comment by JonathanAquino-NextRoll** [comment]
Thank you - hope this can get implemented.

**Comment by JonathanAquino-NextRoll** [comment]
I think this would solved by https://github.com/WICG/attribution-reporting-api/blob/main/aggregate_key_discovery.md, i.e., querying a range of buckets instead of predefined buckets

**Comment by csharrison** [comment]
@JonathanAquino-NextRoll , yes! I think the proposal solves your use-case where you aren't interested in pre-declaring all buckets. We are planning on discussing this in our next meeting on the 26th if you want to attend.

**Comment by KiHunCho** [comment]
Hello,
I’m currently evaluating the adoption of the Attribution Reporting API for our company.

Since "Group By" functionality is very important for reporting, I’ve reviewed the [aggregation_key_discovery explainer](https://github.com/WICG/attribution-reporting-api/blob/main/aggregate_key_discovery.md).
I would like to ask whether it’s possible to perform "Group By" using the key discovery functionality in the Aggregation Service.

For example, 
If the aggregation key is composed of both "geo" and "campaign_id", is it possible to obtain aggregated results based on the sub-bucket (or sub-key) "geo" only?
If I want to aggregate by "geo" only, would I need to define a separate aggregation key for that purpose?

I’d like to clarify the meaning of "non pre-declared bucket matching the key-mask" as described in the algorithm section of the explainer — it’s a bit unclear to me.

Thank you!

**Comment by gh77cho** [comment]
@JonathanAquino-NextRoll Yes, I know that it hasn't been released yet, and I'm hoping that 'Group By' will become possible through key discovery.

I couldn't quite understand the explanation in the explainer about how buckets are matched with a key-mask, so I wanted to know whether the goal is to support 'Group By' through key discovery.

Thank you for your quick response.

**Comment by KiHunCho** [comment]
@JonathanAquino-NextRoll 
I’m sorry if I misunderstood your requirements.
It seems that key-discovery might be able to address your use case.
Thank you!
